stl/inc/atomic

```diff
@@ -1675,6 +1795,101 @@ template <class _Ty>
 atomic(_Ty) -> atomic<_Ty>;
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+struct atomic_ref : _Choose_atomic_base_t<_Ty, _Ty&> { // atomic reference
+private:
+    using _Base = _Choose_atomic_base_t<_Ty, _Ty&>;
+
+public:
+    // clang-format off
+    static_assert(is_trivially_copyable_v<_Ty> && is_copy_constructible_v<_Ty> && is_move_constructible_v<_Ty>
+        && is_copy_assignable_v<_Ty> && is_move_assignable_v<_Ty>,
+        "atomic<T> requires T to be trivially copyable, copy constructible, move constructible, copy assignable, "
+        "and move assignable.");
+    // clang-format on
+
+    using value_type = _Ty;
+
+    explicit atomic_ref(_Ty& _Value) : _Base(_Value) {
+        if constexpr (_Is_potentially_lock_free) {
+            _Check_aligment(_Value);
+        }
+
+#if 1 // TRANSITION, ABI
+        if constexpr (!is_always_lock_free) {
+#else
+        if constexpr (!_Is_potentially_lock_free) {
+#endif
+            this->_Init_spinlock_for_ref();
+        }
+    }
+
+    atomic_ref(const atomic_ref&) noexcept = default;
+
+    atomic_ref& operator=(const atomic_ref&) = delete;
+
+    static constexpr bool is_always_lock_free = _Is_always_lock_free<sizeof(_Ty)>;
+
+    static constexpr bool _Is_potentially_lock_free =

```

Comment:
I think we should do the 'attempt to use cmpxchg16b if it is available' dance here for `atomic_ref<2*sizeof(void)>` since it's a new type; I think I have code for this detection already lying around, stand by...

Reply:
I've revived `atomic_ref<16 bytes>`, and put the functions in atomic ref satellite. Please add this implementation there

Reply:
Actually I think I know what should go there, added